File: codengine-runner-ts/src/runtime.ts
// The codengine IR executor. Implements codengine-spec/semantics/execution.md.
// The runner is "dumb": it trusts the precomputed executionPlan and only resolves
// functions and applies the runtime rules.

import type { Task, WorkflowIR } from "./types.js";

/** A single input/output object flowing between tasks. */
export type TaskData = Record<string, unknown>;
export type TaskFunction = (input: TaskData) => unknown;
export type FunctionMap = Record<string, TaskFunction>;

// Per-task run state:
//   TaskData[]  ran and produced these outputs
//   null        ran but produced no data (a halt, or a router)
//   undefined   never ran (skipped)
type State = TaskData[] | null | undefined;

function euclideanIndex(n: number, length: number): number {
  return ((n % length) + length) % length;
}

// Cartesian product of the fanIn outputs, each combination merged into one object.
function cartesianMerge(outputs: TaskData[][]): TaskData[] {
  return outputs
    .reduce<TaskData[][]>((acc, curr) => acc.flatMap((a) => curr.map((c) => [...a, c])), [[]])
    .map((combination) => Object.assign({}, ...combination));
}

// Apply the `^key` (input) / `key$` (output) rename directives declared in args.
function formatData(task: Task, data: TaskData, type: "input" | "output"): TaskData {
  const directives = Object.keys(task.args).filter((k) => k.startsWith("^") || k.endsWith("$"));
  if (directives.length === 0) return data;

  const result: TaskData = { ...data };
  for (const directive of directives) {
    const isInput = directive.startsWith("^");
    const source = isInput ? directive.slice(1) : directive.slice(0, -1);
    const kind = isInput ? "input" : "output";
    if (kind === type && source in data) {
      const target = task.args[directive] as string;
      result[target] = data[source];
      delete result[source];
    }
    delete result[directive];
  }
  return result;
}

function inject(injected: Map<string, TaskData[]>, target: string, input: TaskData): void {
  const existing = injected.get(target) ?? [];
  existing.push(input);
  injected.set(target, existing);
}

// Handle one function result. Pushes data outputs, or routes (injecting the
// transferred input into the selected target). Returns true if it routed.
function classify(
  task: Task,
  result: unknown,
  input: TaskData,
  outputs: TaskData[],
  injected: Map<string, TaskData[]>,
): boolean {
  // No data (halts the branch). Classify by type, never truthiness: 0, "" and {}
  // are meaningful and handled below.
  if (result === null || result === undefined || result === false) return false;

  if (result === true) {
    // Passthrough: output equals the input.
    outputs.push(formatData(task, input, "output"));
    return false;
  }

  if (typeof result === "string") {
    const route = task.routes.find((r) => r.label === result);
    if (route) inject(injected, route.target, input); // no match -> halt
    return true;
  }

  if (typeof result === "number" && Number.isInteger(result)) {
    if (task.fanOut.length > 0) {
      const target = task.fanOut[euclideanIndex(result, task.fanOut.length)];
      inject(injected, target, input);
    }
    return true;
  }

  if (Array.isArray(result)) {
    for (const item of result) {
      if (item === null || typeof item !== "object" || Array.isArray(item)) {
        throw new Error(`Task '${task.name}' returned an array with a non-object item.`);
      }
      outputs.push(formatData(task, item as TaskData, "output"));
    }
    return false;
  }

  if (typeof result === "object") {
    outputs.push(formatData(task, result as TaskData, "output"));
    return false;
  }

  throw new Error(`Task '${task.name}' returned an unsupported value of type ${typeof result}.`);
}

/**
 * Run a workflow from `entry` with `input`. Returns the `output` task's collected
 * output, or `null` if it never ran.
 */
export function run(
  ir: WorkflowIR,
  functions: FunctionMap,
  entry: string,
  input: TaskData = {},
): TaskData[] | null {
  const tasks = new Map(ir.tasks.map((task) => [task.name, task]));
  const entryTask = tasks.get(entry);
  if (!entryTask) throw new Error(`Unknown entry task '${entry}'.`);

  const state = new Map<string, State>();
  const injected = new Map<string, TaskData[]>();

  for (const name of entryTask.executionPlan) {
    const task = tasks.get(name);
    if (!task) continue;

    let inputs: TaskData[];
    const injectedInputs = injected.get(name);
    if (injectedInputs) {
      inputs = injectedInputs;
    } else {
      // A required fanIn that ran and produced no data blocks this task.
      const blocked = task.fanIn.some(
        (f) => !task.fanInNullable.includes(f) && state.get(f) === null,
      );
      if (blocked) {
        state.set(name, undefined);
        continue;
      }
      const present = task.fanIn
        .map((f) => state.get(f))
        .filter((output): output is TaskData[] => Array.isArray(output));
      if (present.length === 0) {
        if (task.fanIn.length === 0) {
          inputs = [{}]; // root task
        } else {
          state.set(name, undefined); // no producer ran
          continue;
        }
      } else {
        inputs = cartesianMerge(present);
      }
    }

    // The run input replaces the entry task's declared args.
    const args = name === entry ? input : task.args;
    const fn = functions[task.function];
    if (!fn) throw new Error(`No function bound for '${task.function}' (task '${name}').`);

    const outputs: TaskData[] = [];
    let routed = false;
    for (const raw of inputs) {
      const formatted = formatData(task, { ...raw, ...args }, "input");
      routed = classify(task, fn(formatted), formatted, outputs, injected) || routed;
    }

    state.set(name, routed ? null : outputs.length > 0 ? outputs : null);
  }

  const result = state.get("output");
  return result === undefined ? null : result;
}

File: codengine-runner-py/codengine_runner/runtime.py
"""The codengine IR executor, in Python.

Implements codengine-spec/semantics/execution.md and is a faithful port of
codengine-runner-ts. The runner is "dumb": it trusts the precomputed
executionPlan and only resolves functions and applies the runtime rules.
"""

import inspect
from itertools import product
from typing import Any, Callable, Optional

TaskData = dict[str, Any]
TaskFunction = Callable[..., Any]
FunctionMap = dict[str, TaskFunction]


class MissingInputError(Exception):
    """A required task input (a parameter with no default) was not provided."""


_NAMED_KINDS = (
    inspect.Parameter.POSITIONAL_OR_KEYWORD,
    inspect.Parameter.KEYWORD_ONLY,
)


def _invoke(task: dict, fn: TaskFunction, data: TaskData) -> Any:
    """Call `fn` with `data` bound as named arguments (see the invocation contract).

    Passes only the entries the function declares; if it declares `**kwargs`, passes
    everything. A required parameter (no default) with no matching input raises a
    normalized MissingInputError.
    """
    parameters = inspect.signature(fn).parameters
    named = {name: p for name, p in parameters.items() if p.kind in _NAMED_KINDS}

    missing = [
        name
        for name, p in named.items()
        if p.default is inspect.Parameter.empty and name not in data
    ]
    if missing:
        raise MissingInputError(
            f"Task '{task['name']}': missing required input(s): {', '.join(missing)}"
        )

    if any(p.kind is inspect.Parameter.VAR_KEYWORD for p in parameters.values()):
        return fn(**data)
    return fn(**{name: data[name] for name in named if name in data})


def _euclidean_index(n: int, length: int) -> int:
    return ((n % length) + length) % length


def _cartesian_merge(outputs: list[list[TaskData]]) -> list[TaskData]:
    """Cartesian product of the fanIn outputs, each combination merged into one."""
    merged: list[TaskData] = []
    for combination in product(*outputs):
        item: TaskData = {}
        for part in combination:
            item.update(part)
        merged.append(item)
    return merged


def _format_data(task: dict, data: TaskData, kind: str) -> TaskData:
    """Apply the `^key` (input) / `key$` (output) rename directives from args."""
    directives = [k for k in task["args"] if k.startswith("^") or k.endswith("$")]
    if not directives:
        return data

    result = dict(data)
    for directive in directives:
        is_input = directive.startswith("^")
        source = directive[1:] if is_input else directive[:-1]
        directive_kind = "input" if is_input else "output"
        if directive_kind == kind and source in data:
            target = task["args"][directive]
            result[target] = data[source]
            del result[source]
        result.pop(directive, None)
    return result


def _inject(injected: dict[str, list[TaskData]], target: str, data: TaskData) -> None:
    injected.setdefault(target, []).append(data)


def _classify(
    task: dict,
    result: Any,
    task_input: TaskData,
    outputs: list[TaskData],
    injected: dict[str, list[TaskData]],
) -> bool:
    """Handle one function result. Pushes data outputs, or routes (injecting the
    transferred input into the selected target). Returns True if it routed.

    Classification is by type, never truthiness: {}, 0 and "" are meaningful.
    bool is checked before int (in Python bool is a subclass of int).
    """
    if result is None or result is False:
        return False

    if result is True:
        outputs.append(_format_data(task, task_input, "output"))
        return False

    if isinstance(result, str):
        route = next((r for r in task["routes"] if r["label"] == result), None)
        if route is not None:  # no match -> halt
            _inject(injected, route["target"], task_input)
        return True

    if isinstance(result, int) and not isinstance(result, bool):
        fan_out = task["fanOut"]
        if fan_out:
            target = fan_out[_euclidean_index(result, len(fan_out))]
            _inject(injected, target, task_input)
        return True

    if isinstance(result, list):
        for item in result:
            if not isinstance(item, dict):
                raise ValueError(f"Task '{task['name']}' returned a non-object array item.")
            outputs.append(_format_data(task, item, "output"))
        return False

    if isinstance(result, dict):
        outputs.append(_format_data(task, result, "output"))
        return False

    raise ValueError(
        f"Task '{task['name']}' returned an unsupported value of type {type(result).__name__}."
    )


def run(
    ir: dict,
    functions: FunctionMap,
    entry: str,
    data: Optional[TaskData] = None,
) -> Optional[list[TaskData]]:
    """Run a workflow from `entry` with `data`. Returns the `output` task's
    collected output, or None if it never ran.
    """
    run_input: TaskData = data or {}
    tasks = {task["name"]: task for task in ir["tasks"]}
    if entry not in tasks:
        raise ValueError(f"Unknown entry task '{entry}'.")

    # state: name present -> list[dict] (produced) or None (ran, no data).
    # A name absent from `state` means the task never ran (skipped).
    state: dict[str, Optional[list[TaskData]]] = {}
    injected: dict[str, list[TaskData]] = {}

    for name in tasks[entry]["executionPlan"]:
        task = tasks.get(name)
        if task is None:
            continue

        if name in injected:
            inputs = injected[name]
        else:
            # A required fanIn that ran and produced no data (None) blocks this task.
            blocked = any(
                f not in task["fanInNullable"] and f in state and state[f] is None
                for f in task["fanIn"]
            )
            if blocked:
                continue  # skipped: leave absent
            present = [state[f] for f in task["fanIn"] if isinstance(state.get(f), list)]
            if not present:
                if not task["fanIn"]:
                    inputs = [{}]  # root task
                else:
                    continue  # no producer ran
            else:
                inputs = _cartesian_merge(present)

        # The run input replaces the entry task's declared args.
        args = run_input if name == entry else task["args"]
        fn = functions.get(task["function"])
        if fn is None:
            raise ValueError(f"No function bound for '{task['function']}' (task '{name}').")

        outputs: list[TaskData] = []
        routed = False
        for raw in inputs:
            formatted = _format_data(task, {**raw, **args}, "input")
            if _classify(task, _invoke(task, fn, formatted), formatted, outputs, injected):
                routed = True

        state[name] = None if routed else (outputs if outputs else None)

    return state.get("output")
